### PR TITLE
kasli: add support for windows platform.

### PR DESCRIPTION
**Pull request by KaifengC**
# ARTIQ Pull Request

## Description of Changes

Add Windows platform support to kasli_tester.py

### Related Issue

None

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :bug: Bug fix  |

### All Pull Requests

- [x] Use correct spelling and grammar.
- [x] Check the copyright situation of your changes and sign off your patches (`git commit --signoff`, see [copyright](../CONTRIBUTING.rst#copyright-and-sign-off)).

### Code Changes

- [x] Run `flake8` to check code style (follow PEP-8 style). `flake8` has issues with parsing Migen/gateware code, ignore as necessary.

### Documentation Changes

None

### Git Logistics

- [x] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

### CODE REVIEW

File: artiq/examples/kasli_basic/repository/kasli_tester.py [hunk 1]
```diff
@@ -1,4 +1,6 @@
 import sys
+if sys.platform == 'win32':
```
Comment (sbourdeauducq) [inline artiq/examples/kasli_basic/repository/kasli_tester.py]: Elsewhere in ARTIQ, this is tested with ``os.name == "nt"``.

File: artiq/examples/kasli_basic/repository/kasli_tester.py [hunk 2]
```diff
@@ -16,11 +18,18 @@ def chunker(seq, size):
...
+            sys.stdin.read(1)
+            return True
+        else:
+            return False        
+    elif sys.platform == 'win32':   
+        if msvcrt.kbhit():
+            sys.stdin.read(1)
+            return True
+        else:
+            return False
```
Comment (sbourdeauducq) [inline artiq/examples/kasli_basic/repository/kasli_tester.py]: The function will break if ``sys.platform`` is neither ``linux2`` or ``win32``.
I would simply test for ``os.name == "nt"`` and use ``msvcrt`` in that case and ``select`` (which is available on more platforms) in all other cases.

File: artiq/examples/kasli_basic/repository/kasli_tester.py [hunk 3]
```diff
@@ -16,11 +18,18 @@ def chunker(seq, size):
...
-        sys.stdin.read(1)
-        return True
-    else:
-        return False
+    if sys.platform == 'linux2':
+        if select.select([sys.stdin,], [], [], 0.0)[0]:
+            sys.stdin.read(1)
+            return True
+        else:
+            return False
```
Comment (sbourdeauducq) [inline artiq/examples/kasli_basic/repository/kasli_tester.py]: This code can be factored:
``
if os.name == "nt":
   is_key_pressed = lambda: select.select([sys.stdin,], [], [], 0.0)[0]
else:
  import msvcrt
  is_key_pressed = msvcrt.kbhit
```

File: artiq/examples/kasli_basic/repository/kasli_tester.py [hunk 4]
```diff
@@ -16,11 +18,18 @@ def chunker(seq, size):
...
-        return False
+    if sys.platform == 'linux2':
+        if select.select([sys.stdin,], [], [], 0.0)[0]:
+            sys.stdin.read(1)
+            return True
+        else:
+            return False        
+    elif sys.platform == 'win32':   
+        if msvcrt.kbhit():
+            sys.stdin.read(1)
```
Comment (sbourdeauducq) [inline artiq/examples/kasli_basic/repository/kasli_tester.py]: Doesn't this block (and break) if the key pressed is not ENTER?

Comment (sbourdeauducq) [reply artiq/examples/kasli_basic/repository/kasli_tester.py]: On Linux, the ``select`` call waits for ENTER, and ``sys.stdin.read`` also waits for ENTER.
On Windows, the ``kbhit`` call does **NOT** wait for ENTER, and I suspect ``sys.stdin.read`` may still wait for ENTER.

Comment (KaifengC) [reply artiq/examples/kasli_basic/repository/kasli_tester.py]: Yes, on windows `kbhit()` will immditely return `0`, which is not `True`. But if you press any button, a none-zero value will return, which is `True`.
Then `sys.stdin.read(1)` will block untill ENTER.

File: artiq/examples/kasli_basic/repository/kasli_tester.py [hunk 5]
```diff
@@ -15,12 +16,18 @@ def chunker(seq, size):
...
-        return False
+if os.name == "nt":
+   import msvcrt
+   is_key_pressed = msvcrt.kbhit
+else:
+   is_key_pressed = lambda: select.select([sys.stdin,], [], [], 0.0)[0]
+        
+def is_enter_pressed() -> TBool: 
+	if is_key_pressed:
+		sys.stdin.read(1)
```
Comment (jordens) [inline artiq/examples/kasli_basic/repository/kasli_tester.py]: Indentation/tabs/spaces. Please run a linter (e.g. `flake8`).

File: artiq/examples/kasli_basic/repository/kasli_tester.py (hunk 4 above)
Comment (sbourdeauducq) [reply artiq/examples/kasli_basic/repository/kasli_tester.py]: Then the behavior is incorrect. You may want to call ``msvcrt.getch`` on Windows instead (and test for the ENTER key so it's consistent with Linux).

Comment (sbourdeauducq) [reply artiq/examples/kasli_basic/repository/kasli_tester.py]: The problem is: if the user presses a key other than ENTER and nothing else, then the program freezes.
Characters displayed on the console are not important.

Comment (KaifengC) [reply artiq/examples/kasli_basic/repository/kasli_tester.py]: > On Linux, the `select` call waits for ENTER, and `sys.stdin.read` also waits for ENTER.

Just want to clarify that on linux, did the `sys.stdin.read(1)` get the last line from `select` or it started wait for a new line?
It sounds like I need to press ENTER twice to `return True`

Comment (sbourdeauducq) [reply artiq/examples/kasli_basic/repository/kasli_tester.py]: ``select`` does not consume any characters, it only checks if something can be read.
So, pressing once is sufficient.

File: artiq/examples/kasli_basic/repository/kasli_tester.py [hunk 6]
```diff
@@ -16,18 +19,18 @@ def chunker(seq, size):
...
-   is_key_pressed = lambda: select.select([sys.stdin,], [], [], 0.0)[0]
-        
 def is_enter_pressed() -> TBool: 
-	if is_key_pressed:
-		sys.stdin.read(1)
-		return True
-	else:
-		return False
+    if os.name == 'nt':
+        if msvcrt.getch() == b'\r':
```
Comment (sbourdeauducq) [inline artiq/examples/kasli_basic/repository/kasli_tester.py]: Nope. This one will block. You need to call kbhit and then getch.

**Comment by sbourdeauducq** [comment]
Thanks!

**Pull request merged**